src/metaswitch/ellis/prov_tools/utils.py

```diff
@@ -236,3 +244,48 @@ def display_user(private_id, public_id, short=False):
             success = False
 
     return success
+
+
+def list_users(target_users_per_chunk=100, keep_going=False):
+    MAX_CHUNK_PROPORTION = 2**24
+    chunk_proportion = MAX_CHUNK_PROPORTION
+    chunk = 0
+    exception = StopIteration
+    while chunk < chunk_proportion:
+        try:
+            _log.debug("Bulk-retrieving public IDs (chunk %d/%d)\n", chunk, chunk_proportion)
+            public_ids = get_users(chunk, chunk_proportion)
+            _log.debug("Retrieved %d public IDs\n", len(public_ids))
+            for public_id in (public_id_obj['public_id'] for public_id_obj in public_ids):
+                yield public_id
+
+            # Move onto the next chunk and then decide whether we should change chunk size.
+            chunk += 1
+            if len(public_ids) < target_users_per_chunk * 0.5 and chunk % 2 == 0:
+                _log.debug("Too few public IDs in chunk - increase chunk size")
+                chunk /= 2
+                chunk_proportion /= 2
+            elif len(public_ids) > target_users_per_chunk * 2 and chunk_proportion < MAX_CHUNK_PROPORTION:
```

Comment:
Shouldn't you check that `chunk_proportion * 2 < MAX_CHUNK_PROPORTION`? Otherwise you'll go over.